From 469f1541a3a2652029f24376732a13ca6aabf6d8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lasse=20Brudeskar=20Vik=C3=A5s?= <bvx89.at.cs@gmail.com>
Date: Sun, 19 Jan 2014 22:01:57 +0100
Subject: [PATCH 1/2] DownloadProvider: Display download speed in notification

Add the current total download speed to in-progress downloads
shown in the notification pane.

Change-Id: I801dbe61c7ee59d0c1d14d5851ad6dc3a7678499
---
 res/values/cm_plurals.xml                     | 33 +++++++++++++++
 res/values/cm_strings.xml                     | 20 +++++++++
 .../providers/downloads/DownloadNotifier.java | 42 ++++++++++++++++---
 3 files changed, 89 insertions(+), 6 deletions(-)
 create mode 100644 res/values/cm_plurals.xml
 create mode 100644 res/values/cm_strings.xml

diff --git a/res/values/cm_plurals.xml b/res/values/cm_plurals.xml
new file mode 100644
index 00000000..b201ff51
--- /dev/null
+++ b/res/values/cm_plurals.xml
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     Copyright (C) 2014 The CyanogenMod Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
+    <!-- Phrase describing a time duration using seconds [CHAR LIMIT=16] -->
+    <plurals name="duration_seconds">
+        <item quantity="one"><xliff:g id="count">%d</xliff:g> sec</item>
+        <item quantity="other"><xliff:g id="count">%d</xliff:g> secs</item>
+    </plurals>
+    <!-- Phrase describing a time duration using minutes [CHAR LIMIT=16] -->
+    <plurals name="duration_minutes">
+        <item quantity="one"><xliff:g id="count">%d</xliff:g> min</item>
+        <item quantity="other"><xliff:g id="count">%d</xliff:g> mins</item>
+    </plurals>
+    <!-- Phrase describing a time duration using hours [CHAR LIMIT=16] -->
+    <plurals name="duration_hours">
+        <item quantity="one"><xliff:g id="count">%d</xliff:g> hr</item>
+        <item quantity="other"><xliff:g id="count">%d</xliff:g> hrs</item>
+    </plurals>
+</resources>
diff --git a/res/values/cm_strings.xml b/res/values/cm_strings.xml
new file mode 100644
index 00000000..9fca450b
--- /dev/null
+++ b/res/values/cm_strings.xml
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     Copyright (C) 2013-2016 The CyanogenMod Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
+    <!-- Notification text containing download duration left and download speed -->
+    <string name="text_download_speed"><xliff:g id="text">%1$s</xliff:g>, <xliff:g id="size" example="230 kB">%2$s</xliff:g>/s</string>
+</resources>
diff --git a/src/com/android/providers/downloads/DownloadNotifier.java b/src/com/android/providers/downloads/DownloadNotifier.java
index d38aa755..e6a9864a 100644
--- a/src/com/android/providers/downloads/DownloadNotifier.java
+++ b/src/com/android/providers/downloads/DownloadNotifier.java
@@ -40,6 +40,7 @@ import android.provider.Downloads;
 import android.service.notification.StatusBarNotification;
 import android.text.TextUtils;
 import android.text.format.DateUtils;
+import android.text.format.Formatter;
 import android.util.ArrayMap;
 import android.util.IntArray;
 import android.util.Log;
@@ -285,6 +286,7 @@ public class DownloadNotifier {
             // Calculate and show progress
             String remainingText = null;
             String percentText = null;
+            String speedAsSizeText = null;
             if (type == TYPE_ACTIVE) {
                 long current = 0;
                 long total = 0;
@@ -311,8 +313,26 @@ public class DownloadNotifier {
 
                     if (speed > 0) {
                         final long remainingMillis = ((total - current) * 1000) / speed;
+                        final int duration, durationResId;
+
+                        // This duplicates DateUtils.formatDuration(),
+                        // but uses our abbreviated plurals.
+                        if (remainingMillis >= DateUtils.HOUR_IN_MILLIS) {
+                            duration = (int) ((remainingMillis + 1800000)
+                                    / DateUtils.HOUR_IN_MILLIS);
+                            durationResId = R.plurals.duration_hours;
+                        } else if (remainingMillis >= DateUtils.MINUTE_IN_MILLIS) {
+                            duration = (int) ((remainingMillis + 30000)
+                                    / DateUtils.MINUTE_IN_MILLIS);
+                            durationResId = R.plurals.duration_minutes;
+                        } else {
+                            duration = (int) ((remainingMillis + 500)
+                                    / DateUtils.SECOND_IN_MILLIS);
+                            durationResId = R.plurals.duration_seconds;
+                        }
                         remainingText = res.getString(R.string.download_remaining,
-                                DateUtils.formatDuration(remainingMillis));
+                                res.getQuantityString(durationResId, duration, duration));
+                        speedAsSizeText = Formatter.formatFileSize(mContext, speed);
                     }
 
                     final int percent = (int) ((current * 100) / total);
@@ -329,11 +349,16 @@ public class DownloadNotifier {
                 builder.setContentTitle(getDownloadTitle(res, cursor));
 
                 if (type == TYPE_ACTIVE) {
-                    final String description = cursor.getString(UpdateQuery.DESCRIPTION);
-                    if (!TextUtils.isEmpty(description)) {
-                        builder.setContentText(description);
+                    if (speedAsSizeText != null) {
+                        builder.setContentText(res.getString(R.string.text_download_speed,
+                                remainingText, speedAsSizeText));
                     } else {
-                        builder.setContentText(remainingText);
+                        final String description = cursor.getString(UpdateQuery.DESCRIPTION);
+                        if (!TextUtils.isEmpty(description)) {
+                            builder.setContentText(description);
+                        } else {
+                            builder.setContentText(remainingText);
+                        }
                     }
                     builder.setContentInfo(percentText);
 
@@ -366,7 +391,12 @@ public class DownloadNotifier {
                             R.plurals.notif_summary_active, cluster.size(), cluster.size()));
                     builder.setContentText(remainingText);
                     builder.setContentInfo(percentText);
-                    inboxStyle.setSummaryText(remainingText);
+                    if (speedAsSizeText != null) {
+                        inboxStyle.setSummaryText(res.getString(R.string.text_download_speed,
+                                remainingText, speedAsSizeText));
+                    } else {
+                        inboxStyle.setSummaryText(remainingText);
+                    }
 
                 } else if (type == TYPE_WAITING) {
                     builder.setContentTitle(res.getQuantityString(
-- 
2.17.1

